### ISSUE TITLE
preprocessing: filter rare events before adding timeline delta tokens

### ISSUE THREAD

**Issue by mmcdermott** [mmcdermott/MEDS_EIC_AR#126]
## The bug

`src/MEDS_EIC_AR/preprocessing/configs/_data.yaml` (and its `_reshard_data.yaml` twin) run MTD stages in this order:

1. `add_time_derived_measurements` — inserts `TIMELINE//DELTA//<unit>` tokens between every consecutive pair of events, carrying the elapsed time as a numeric value.
2. `count_codes` — aggregates `code/n_subjects`.
3. `filter_measurements` — drops codes with `n_subjects < MIN_SUBJECTS_PER_CODE` (preserves `TIMELINE/*`, `MEDS_DEATH/*`, `MEDS_BIRTH/*`, admission/discharge/registration via the regex match-revise).
4. `filter_subjects` — drops subjects with too few events.
5. `fit_quantile_binning` + `bin_numeric_values` — quantize the delta values.

Because the delta tokens are computed in step (1) on the **unfiltered** event stream but the rare events between them are dropped in step (3), a subject's kept-events sequence can contain **two or more consecutive `TIMELINE//DELTA/*` tokens** with nothing between them. The model sees:

```
DELTA(1h)  →  <rare_event_dropped_in_step_3>  →  DELTA(30m)  →  kept_event
```

becoming

```
DELTA(1h)  →  DELTA(30m)  →  kept_event
```

after `filter_measurements`.

## Why it matters

For an autoregressive "everything-is-code" model, training data is a sequence of tokens. Back-to-back delta tokens teach the model a noisy pattern — "a delta can be followed by another delta" — that doesn't reflect any real-world semantics, because in a world where the model only knows kept codes, a delta should always be followed by some clinical event (or an EOS / END-OF-TIMELINE sentinel). The pattern pollutes:

- **The transition distribution**: "what code typically follows a delta?" picks up spurious mass on delta itself.
- **The quantile binning** (step 5): the quantiles over `TIMELINE//DELTA//years` values include the deltas whose neighbors were about to be dropped, skewing the bin boundaries toward shorter-than-clinically-meaningful intervals.
- **Generation quality**: the model learns to occasionally emit a delta immediately after a delta, which downstream `format_trajectories` has to deal with. Not a correctness bug — the generated parquets are valid — but it's a lower-quality output distribution than we should be producing.

The magnitude depends on `MIN_SUBJECTS_PER_CODE`. For small thresholds (~100) on MIMIC-IV scale, probably a few percent of training-sequence positions. For large thresholds (1000+, as the default for real runs), it can be a sizeable chunk of the filtered-out mass.

## Fix

**Reorder the pipeline so filtering happens before time-derived-token insertion.**

Proposed new order:

1. `count_codes` (on raw events, no timeline tokens yet — fine, the regex in `filter_measurements` still covers MEDS_DEATH/MEDS_BIRTH/admission/discharge/registration; the `.*TIME.*` match becomes dead code but is harmless).
2. `filter_measurements`
3. `filter_subjects`
4. `add_time_derived_measurements` — now operates on the filtered event stream; each delta truly represents time between *kept* events.
5. `fit_quantile_binning`
6. `bin_numeric_values`

This gives each kept event exactly one preceding delta — clean token sequence, clean quantiles.

## Tradeoffs to flag explicitly

**Delta semantics change.** Under the current ordering, a delta = "time since previous raw event" (some of which we then drop). Under the new ordering, a delta = "time since previous **kept** event". Consequences:

- **Values get larger.** A kept event at 12:00 preceded by a dropped event at 11:30 and another kept event at 10:00 currently produces `DELTA(30m)` preceding the kept event (and a bonus `DELTA(1h30m)` floating before the dropped event). Under the fix, the kept event gets `DELTA(2h)`.
- **Quantile bin distribution shifts right.** The `custom_bins` in `_data.yaml` (1_minute through 40_years) span the relevant range, so this is just a redistribution of mass among existing bins, not a coverage issue.
- **Anyone comparing training runs before/after the fix** will see measurably different loss curves because the vocab composition shifts. Existing pretrained checkpoints are not invalidated in the weights-are-garbage sense, but the quantile-bin interpretation they were trained against no longer matches the preprocessing output.

**Is this the *right* semantics?** For an autoregressive language-model-over-MEDS-codes, yes — the model is conditioning on the preceding *observed* tokens, so "time since previous observed" is exactly what the delta should encode. The other interpretation (time since last raw event) would only be right if the model had to predict dropped events, which by construction it doesn't.

## Scope

In:
- Reorder stages in `_data.yaml` and `_reshard_data.yaml`.
- Regenerate doctests whose expected outputs encode the old quantile distribution (likely `tests/grammar/test_in_process.py` / `tests/grammar/test_cli.py` if any of them depend on specific TIMELINE//DELTA bin populations — spot-check before re-regenerating).
- Add a test that ingests a fixture with deliberately-rare codes and asserts no two consecutive `TIMELINE//DELTA/*` tokens appear in the processed output.

Out:
- Migrating existing pretrained checkpoints. This is a data-preprocessing semantics change; old checkpoints stay old, new runs get the corrected pipeline. Frame this as a one-way door in the PR.
- Changing the `custom_bins` values. The existing bins still span the relevant range under the new semantics; tuning them is its own follow-up.

## Priority

Medium. Not a correctness-breaking bug (the output is valid MEDS, generation still works, grammar tests still pass), but it's a real training-data-quality issue that will compound over many training runs. Worth fixing before the next major training sweep.

cc #108 (Llama swap PR discussed training-from-scratch; this is a good moment to also fix the data-side quality issue before anyone does a long training run).
